nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix

```diff
@@ -148,15 +232,61 @@ in {
     assertions = [
       {
         assertion = (config.boot.kernelPackages.kernel.features or { efiBootStub = true; }) ? efiBootStub;
-
         message = "This kernel does not support the EFI boot stub";
       }
+      {
+        assertion = builtins.all (filename: !(hasInfix "/" filename)) (builtins.attrNames cfg.extraEntries);
+        message = "systemd-boot does not support entries within folders";
+      }
+      {
+        assertion = builtins.all (hasSuffix ".conf") (builtins.attrNames cfg.extraEntries);
+        message = "All systemd-boot entries must have a .conf file extension";
+      }
+      {
+        assertion = builtins.all (filename: !(hasPrefix "/" filename)) (builtins.attrNames cfg.extraFiles);
+        message = "Paths to extra files must not begin with a slash";
+      }
+      {
+        assertion = builtins.all (filename: !(hasInfix ".." filename)) (builtins.attrNames cfg.extraFiles);
+        message = "Paths to extra files must not reference the parent directory";
+      }
+      {
+        assertion = builtins.all (filename: !(hasInfix "nixos/.extra-files" filename)) (map toLower (builtins.attrNames cfg.extraFiles));
+        message = "Files cannot be placed in the nixos/.extra-files directory";
+      }
     ];
 
     boot.loader.grub.enable = mkDefault false;
 
     boot.loader.supportsInitrdSecrets = true;
 
+    boot.loader.systemd-boot.extraFiles = mkMerge [
+      # TODO: This is hard-coded to use the 64-bit EFI app, but it could probably
+      # be updated to use the 32-bit EFI app on 32-bit systems.  The 32-bit EFI
+      # app filename is BOOTIA32.efi.
```

Comment:
you could check `stdenv.targetPlatform.is32bit` for this. maybe it's also worth adding an assert that `stdenv.targetPlatform.isx86` when x86-only items are enabled?

Reply:
This was taken straight from the old code, I didn't make any changes to the `memtest86` stuff as I don't use it

Reply:
yup, did see it. was merely a suggestion for if you felt like addressing that one too while you're at it. :)